# Final Assignment of Capstone Week 3--Segmenting and clustering of neighborhoods of Toronto

## 1. Load and clean data

First load the data from https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
And then clean the data as below:
    *  The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
    *  Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
    *  More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.
    *  If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [8]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
#from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#import requests # library to handle requests
#from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
#import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [9]:

#read_html can load the table using a url;
url="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
df=pd.read_html(url)[0]
df.shape

(180, 3)

In [10]:
#delete the rows with 'Not assigned' borough;
df=df[df['Borough']!='Not assigned']
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
8           M9A         Etobicoke   
9           M1B       Scarborough   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
17          M9B         Etobicoke   
18          M1C       Scarborough   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
26          M9C         Etobicoke   
27          M1E       Scarborough   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
36          M1G       Scarborough   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N      Toronto/York   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
107         M9P         Etobicoke   
108         M1R       Scarborough   
109         M2R        North York   
111         M4R   Central Toronto   
112         M5R   Central Toronto   
113         M6R      West Toronto   
114         M7R       Mississauga   
116         M9R         Etobicoke   
117         M1S       Scarborough   
120         M4S   Central Toronto   
121         M5S  Downtown Toronto   
122         M6S      West Toronto   
126         M1T       Scarborough   
129         M4T   Central Toronto   
130         M5T  Downtown Toronto   
135         M1V       Scarborough   
138         M4V   Central Toronto   
139         M5V  Downtown Toronto   
142         M8V         Etobicoke   
143         M9V         Etobicoke   
144         M1W       Scarborough   
147         M4W  Downtown Toronto   
148         M5W  Downtown Toronto   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Vic

We can see that there are 103 boroughs left after the 'not assigned' borough are deleted from the dataframe.
We also want to double check if the current postal codes are all unique, or in other words, if there are duplicated codes.

In [11]:
df['Postal Code'].unique().size

103

Good. All the 103 postal codes in the table are unique.
Now let's assign the neighourhood name the same as borough name if it's not assigned.

In [12]:
for index, row in df.iterrows(): 
    if row['Neighbourhood']=='Not assigned':
        row['Neighbourhood']=row['Borough']
        

df[df['Neighbourhood']=='Not assigned'].shape
#make sure there is not a 'not assigned' neighorhood now.

(0, 3)

In [13]:
#get the size of the table again
df.shape

(103, 3)

## 2. Get the latitude ang longitude of each neighourhood

First, I tried to get the latitude and longitude  by using geocoder as demonstrated in course guide.

In [10]:
%pip install geocoder
import geocoder # import geocoder


#set the initial values of latitude and longitude to be None;
df['latitude']=None
df['longitude']=None
df


     |████████████████████████████████| 98 kB 1.3 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood latitude longitude  
2                                            Parkwoods     None      None  
3                                     Victoria Village     None      None  
4                            Regent Park, Harbourfront     None      None  
5                     Lawrence Manor, Lawrence Heights     None      None  
6          Queen's Park, Ontario Provincial Government     None      None  
..                                                 ...      ...       ...  
160      The Kingsway, Montgomery Road, Old Mill North     None      None  
165                               Church and Wellesley     None      None  
168  Business reply mail Processing Centre, South C...     None      None  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...     None      None  
178  Mimico NW, The Queensway West, South of Bloor,...     None      None  

[103 rows x 5 columns]

I set up a while loop like below to get the latitude and longitude for one neighbourhood; but it run for more than a minute and still didn't stop.

In [ ]:

lat_lng_coords=None
while (lat_lng_coords is None):
    g = geocoder.google('M7A, Toronto, Ontario')
    lat_lng_coords = g.latlng

lat_lgn_coords

So I decided to go ehead and use the table provided elsewhere.



In [15]:
df_coor=pd.read_csv('https://cocl.us/Geospatial_data')
df_coor.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
#merge the two dataframes together;
df_toronto=pd.merge(df,df_coor,on='Postal Code')
df_toronto.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## 3. Cluster neighbourhood using K-means

Define Foursquare credential and version

In [17]:
CLIENT_ID = 'KTH12FDY3PFVT1TVVGICGZ1EAG0VI3EPPDG0BXU1NER0R232' # your Foursquare ID
CLIENT_SECRET = 'FWNXW4CB3EHEQQFUJN0CLRPNUIUIQEWIGCJHSYWQRGG1JNMX' # your Foursquare Secret
ACCESS_TOKEN = 'AFYA3MZXBJ0MAQ5JFQLJ44GEGZAAI0R5COVXFFKEIFYN2O4M' # your FourSquare Access Token

VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value



In [23]:
import requests
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    return(nearby_venues)

In [24]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )
toronto_venues.head()

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3         Parkwoods               43.753259               -79.329656   
4  Victoria Village               43.725882               -79.315572   

                               Venue  Venue Latitude  Venue Longitude  \
0                    Brookbanks Park       43.751976       -79.332140   
1                    Brookbanks Pool       43.751389       -79.332184   
2                      Variety Store       43.751974       -79.333114   
3  Corrosion Service Company Limited       43.752432       -79.334661   
4             Victoria Village Arena       43.723481       -79.315635   

               Venue Category  
0                        Park  
1                        Pool  
2           Food & Drink Shop  
3  Construction & Landscaping  
4                Hockey Arena

Let's check how many venues were returned for each neighborhood

In [26]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                                5   
Alderwood, Long Branch                                                  10   
Bathurst Manor, Wilson Heights, Downsview North                         22   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       23   
Berczy Park                                                             59   
Birch Cliff, Cliffside West                                              4   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   13   
Cedarbrae                                                                8   
Central Bay Street                                                      60   
Christie                                                                17   
Church and Wellesley                                                    79   
Clarks Corners, Tam O'Shanter, Sullivan                                 12   
Cliffside, Cliffcrest, Scarborough Village West                          4   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                        10   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                         4   
Don Mills                                                               26   
Dorset Park, Wexford Heights, Scarborough Town ...                       7   
Downsview                                                               17   
Dufferin, Dovercourt Village                                            16   
East Toronto, Broadview North (Old East York)                            2   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                       9   
Fairview, Henry Farm, Oriole                                            63   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          4   
Garden District, Ryerson                                               100   
Glencairn                                                                6   
Golden Mile, Clairlea, Oakridge                                          9   
Guildwood, Morningside, West Hill                                        8   
Harbourfront East, Union Station, Toronto Islands                      100   
High Park, The Junction South                                           24   
Hillcrest Village                                                        5   
Humber Summit                                                            3   
Humberlea, Emery                                                         2   
Humewood-Cedarvale                                                       4   
India Bazaar, The Beaches West                                          20   
Kennedy Park, Ionview, East Birchmount Park                              6   
Kensington Market, Chinatown, Grange Park                               61   
Kingsview Village, St. Phillips, Martin Grove G...                       2   
Lawrence Manor, Lawrence Heights                                        12   
Lawrence Park                                                            3   
Leaside                                                                 35   
Little Portugal, Trinity                                                42   
Malvern, Rouge                                                           1

Let's find out how many unique categories can be curated from all the returned venues

In [27]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 274 uniques categories.


In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3         Parkwoods                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0               0                 0                   0      0   
1               0                 0                   0      0   
2               0                 0                   0      0   
3               0                 0                   0      0   
4               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  Bus Station  \
0                0             0              0         0            0   
1                0             0 

we can examine the size of the new dataframe. The column number should be 1 plus the number of venue type (275).

In [31]:
toronto_onehot.shape

(2124, 275)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [35]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()


Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0             0.0      0.0                 0.0             0.0   
1             0.0      0.0                 0.0             0.0   
2             0.0      0.0                 0.0             0.0   
3             0.0      0.0                 0.0             0.0   
4             0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0              0.0               0.0             0.000000           0.0   
1              0.0               0.0             0.000000           0.0   
2              0.0               0.0             0.000000           0.0   
3              0.0               0.0             0.000000           0.0   
4              0.0               0.0             0.043478           0.0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0.0          0.0         0.0                  0.0               0.0   
1       0.0          0.0         0.0                  0.0               0.0   
2       0.0          0.0         0.0                  0.0               0.0   
3       0.0          0.0         0.0                  0.0               0.0   
4       0.0          0.0         0.0                  0.0               0.0   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Baby Store  \
0                 0.0          0.0            0.0        0.0         0.0   
1                 0.1          0.0            0.0        0.0         0.0   
2                 0.0          0.0            0.0        0.0         0.0   
3                 0.0          0.0            0.0        0.0         0.0   
4                 0.0          0.0            0.0        0.0         0.0   

   Bagel Shop  Bakery      Bank  Bar  Baseball Field  Baseball Stadium  \
0         0.0     0.0  0.000000  0.0             0.0               0.0   
1         0.0     0.0  0.000000  0.0             0.0               0.0   
2         0.0     0.0  0.090909  0.0             0.0               0.0   
3         0.0     0.0  0.250000  0.0             0.0               0.0   
4         0.0     0.0  0.000000  0.0             0.0               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0         0.0   
1                 0.0    0.0              0.0       0.0         0.0   
2                 0.0    0.0              0.0       0.0         0.0   
3                 0.0    0.0              0.0       0.0         0.0   
4                 0.0    0.0              0.0       0.0         0.0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0                 0.0        0.0     0.0         0.0            0.0   
1                 0.0        0.0     0.0         0.0            0.0   
2                 0.0        0.0     0.0         0.0            0.0   
3                 0.0        0.0     0.0         0.0            0.0   
4                 0.0        0.0     0.0         0.0            0.0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0        0.0       0.0                   0.0        0.200000      0.0   
1        0.0       0.0                   0.0        0.000000      0.0   
2        0.0       0.0                   0.0        0.000000      0.0   
3        0.0       0.0                   0.0        0.000000      0.0   
4        0.0       0.0                   0.0        0.043478      0.0   

   Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0     

Run k-means to cluster the neighborhood into 5 clusters

In [36]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Now let's create the new dataframe and display the top 10 venues for each neighborhood, and create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt        Clothing Store   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink                Lounge   
1                        Gym           Coffee Shop          Skating Rink   
2                Coffee Shop         Shopping Mall    Frozen Yogurt Shop   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                Coffee Shop        Sandwich Place         Grocery Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot    Falafel Restaurant           Event Space   
1        Sandwich Place    Athletics & Sports                   Pub   
2           Supermarket         Deli / Bodega      Sushi Restaurant   
3   Distribution Center               Dog Run      Doner Restaurant   
4               Butcher             Juice Bar  Fast Food Restaurant   

  8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
0  Ethiopian Restaurant           Escape Room         Distribution Center  
1          Dance Studio              Pharmacy  Construction & Landscaping  
2     Mobile Phone Shop            Restaurant              Ice Cream Shop  
3            Donut Shop             Drugstore                 Yoga Studio  
4        Breakfast Spot          Liquor Store                        Café

Now merge the cluster label with the popular 10 venues of each neighborhood.

In [46]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronto

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.tail() # check the last columns!

Postal Code           Borough  \
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   
99                                Church and Wellesley  43.665860 -79.383160   
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509   
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999   

     Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
98              0.0                 River                  Pool   
99              0.0           Coffee Shop      Sushi Restaurant   
100             0.0           Yoga Studio         Auto Workshop   
101             3.0        Baseball Field           Yoga Studio   
102             0.0        Hardware Store         Tanning Salon   

           3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
98                     Drugstore                 Diner        Discount Store   
99           Japanese Restaurant            Restaurant               Gay Bar   
100                Garden Center                Garden  Fast Food Restaurant   
101  Eastern European Restaurant               Dog Run      Doner Restaurant   
102                Grocery Store            Kids Store  Fast Food Restaurant   

    6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
98    Distribution Center                   Dog Run      Doner Restaurant   
99            Men's Store  Mediterranean Restaurant                  Café   
100        Farmers Market        Light Rail Station            Comic Shop   
101            Donut Shop                 Drugstore   Dumpling Restaurant   
102        Discount Store         Convenience Store        Sandwich Place   

    9th Most Common Venue 10th Most Common Venue  
98             Donut Shop    Dumpling Restaurant  
99            Yoga Studio                  Hotel  
100                  Park            Pizza Place  
101     Electronics Store         Discount Store  
102         Burrito Place           Burger Joint